packages/@react-aria/collections/src/Document.ts

```diff
@@ -412,10 +422,22 @@ export class Document<T, C extends BaseCollection<T> = BaseCollection<T>> extend
     }
 
     this.updateCollection();
+
+    // Reset queuedRender to false when getCollection is called during render.
+    if (!this.inSubscription) {
+      this.queuedRender = false;
+    }
+
     return this.collection;
   }
 
   updateCollection() {
+    // First, update the indices of dirty element children.
+    for (let element of this.dirtyNodes) {
+      element.updateChildIndices();
+    }
+
+    // Next, update dirty collection nodes.
```

Comment:
do in the same loop as above?

Reply:
They need to be separate passes because `dirtyNode` is not topologically ordered. We need to update the child indices first otherwise we could reach a dirty element, update its node, and later update the element's index. That would result in the node having the wrong index (nodes are immutable).